vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution_unblinded.cc

```diff
@@ -201,9 +245,93 @@ void Unblinded::PrepareTokens(
     return;
   }
 
-  // TODO(https://github.com/brave/brave-browser/issues/8887):
-  // we should reserve this tokens and add step STEP_RESERVE
-  PreparePublishers(token_list, std::move(contribution), types, callback);
+  const std::string contribution_id = contribution->contribution_id;
+  const std::string contribution_string =
+      braveledger_bind_util::FromContributionToString(std::move(contribution));
+
+  std::vector<std::string> token_id_list;
+  for (const auto& item : token_list) {
+    token_id_list.push_back(base::NumberToString(item.id));
+  }
+
+  auto reserved_callback = std::bind(
+      &Unblinded::OnMarkUnblindedTokensAsReserved,
+      this,
+      _1,
+      std::move(token_list),
+      contribution_string,
+      types,
+      callback);
+
+  ledger_->MarkUnblindedTokensAsReserved(
+      token_id_list,
+      contribution_id,
+      reserved_callback);
+}
+
+void Unblinded::ReserveStepSaved(
+    const ledger::Result result,
+    const std::vector<ledger::UnblindedToken>& list,
+    const std::string& contribution_string,
+    const std::vector<ledger::CredsBatchType>& types,
+    ledger::ResultCallback callback) {
+  if (result != ledger::Result::LEDGER_OK) {
+    BLOG(0, "Reserve step was not saved");
+    callback(ledger::Result::RETRY);
+    return;
+  }
+
+  auto contribution =
+      braveledger_bind_util::FromStringToContribution(contribution_string);
+  if (!contribution) {
+    BLOG(0, "Contribution was not converted successfully");
+    callback(ledger::Result::RETRY);
+    return;
+  }
+
+  // Discard cached reserved tokens now that retry window has passed
+  DiscardReservedUnblindedTokensFromCache(contribution->contribution_id);
+
+  PreparePublishers(list, std::move(contribution), types, callback);
+}
+
+void Unblinded::OnMarkUnblindedTokensAsReserved(
```

Comment:
to reduce potential flow interruption you could  as part of `MarkUnblindedTokensAsReserved` db query also set contribution to `STEP_RESERVE` so that is done in one transaction. This way it's less error prone. And you can remove caching as well

Reply:
Good idea, done.